File: services/collector/src/dw_collector/resetweek.py
"""Game week boundary: Monday 02:00 UTC.

Implemented three times (SQL reset_week_start, this module, TypeScript); all
three consume protocol-fixtures/reset-week/vectors.json — change together.
"""

from __future__ import annotations

from datetime import UTC, datetime, time, timedelta

RESET_HOUR_UTC = 2


def reset_week_start(ts: datetime) -> datetime:
    """Most recent Monday 02:00 UTC at or before ts (boundary inclusive)."""
    if ts.tzinfo is None:
        msg = "reset_week_start requires a timezone-aware datetime"
        raise ValueError(msg)
    shifted = ts.astimezone(UTC) - timedelta(hours=RESET_HOUR_UTC)
    monday = shifted.date() - timedelta(days=shifted.weekday())
    return datetime.combine(monday, time(hour=RESET_HOUR_UTC), tzinfo=UTC)

File: apps/dashboard/src/lib/resetWeek.ts
// Game week boundary: Monday 02:00 UTC. Implemented three times (SQL,
// Python, this file); all three consume
// protocol-fixtures/reset-week/vectors.json — change them together.

const RESET_HOUR_UTC = 2;
const HOUR_MS = 3_600_000;

/** Most recent Monday 02:00 UTC at or before ts (boundary inclusive). */
export function resetWeekStart(ts: Date): Date {
  const shifted = new Date(ts.getTime() - RESET_HOUR_UTC * HOUR_MS);
  const mondayOffset = (shifted.getUTCDay() + 6) % 7;
  return new Date(
    Date.UTC(
      shifted.getUTCFullYear(),
      shifted.getUTCMonth(),
      shifted.getUTCDate() - mondayOffset,
      RESET_HOUR_UTC,
    ),
  );
}
